### ISSUE TITLE
Unfortunate, almost unusable performance regression in v1.6.1 to 1.7.0

### ISSUE THREAD

**Issue by whorfin** [gerbera/gerbera#1272]
I have a relatively large collection of audio files, approx 250K of them
The database for v1.6.1 was a little over 1G
Gerbera 1.7.0 successfully "migrated" the database, but has been consuming 100% of one core and occasionally emitting messages about "Taglib: can't open mp3 file" for untagged mp3 files.  These messages haven't yet repeated, and seem to indicate some very unperformant passing over every single file, smelling like an exponential or other bad-big-O algorithm.
For three days and counting it's been doing this; the database file is very very slowly growing, like a few kilobytes every ten seconds
I'm about to give up, but figured I'd see if it ever finishes.
Is this expected?  This isn't even a "Rescan from scratch" situation, it is a post-migration.

I can not yet tell if there is an additional regression or the massive CPU usage and a threading problem causing query latency, but in addition, at least at the moment, this behavior completely breaks `foobar2000` on iOS.  When top-level `Albums` or `PC Directory` queries are made, `8player` on iOS seems to do them in chunks.  `foobar2000` seems to request "all the things".  Gerbera, in turn, seems to never return "all the things", either because it's stuck because some re-indexing/updating process is still happening and it never finishes the query, or because it's too damn slow - i've given up watching `foobar2000` be stuck on "Please wait..."

**Comment by KarlStraussberger** [comment]
I observed the same on the first run but somehow it's gone with the next start of the service.

We had a similar discussion in #1077. So the main question is whether you use sqlite or mysql.

**Comment by whorfin** [comment]
sqlite

**Comment by KarlStraussberger** [comment]
I tried to remember the changes I made inbetween. Which version of libpupnp are you using?

**Comment by whorfin** [comment]
1.14.0
i restarted the service and it is back to using 100% of one core while trickling in messages about can't-open-file which really mean missing-tags
Also - that's bad news, as I can see from the logs that the TagLibHandler spew is on files which it already complained about.  1.6.1 did not re-search all missing tag files on server restart

**Comment by whorfin** [comment]
...this doesn't seem like #1077 other than "horribly slow" - one core is pegged, but i'm not seeing crazy write saturation or anything other than the expected reads; except there's more reads than there should be, as the music library hasn't changed but gerbera going over everything again

On the off chance something is just broken badly with database migration, i just nuked the database and will see what happens with 1.7.0 attempting to populate from scratch
Weirdly, CPU usage is lower with total rebuilding than it was with ... whatever it was doing before

**Comment by KarlStraussberger** [comment]
> 1.14.0
> i restarted the service and it is back to using 100% of one core while trickling in messages about can't-open-file which really mean missing-tags
> Also - that's bad news, as I can see from the logs that the TagLibHandler spew is on files which it already complained about in the previous server run. 1.6.1 did not re-search all missing tag files on server restart

- 1.6.1 never came back to the files because of an error in timestamp handling (it only compared with autoscan root instead of the whole tree). We do not know where scanning stopped, so all files have to be scanned again.
- Can you try with UPnP head (1.16)?

**Comment by whorfin** [comment]
While my database is rebuilding...
I had made a database backup, so I can declare that 1.16 (built from head UPnP) didn't seem to matter - still looping over previously-visited files, and the foobar2000 "waits forever" issue is still an issue

**Comment by whorfin** [comment]
Database rebuild is not finished
But if i restart the server, i see TagLibHandler messages that were already omitted
weirdly not all of them, but many repeats
something is not right

**Comment by KarlStraussberger** [comment]
As said, if the scan was not completed (timestamp is set in database) the whole tree needs to be rescanned because we do not know where it stopped. So that's as expected - it did not work correctly since ages (only if scan finished before restart). Once it's done, restart takes a minute or two to set up inotify watch for the whole tree.

Obviously, if the scan is still ongoing, any client access is concurring with database updates. So, if you try to browse the client needs a lot of data and creates many read tasks which are all handled sequentially.

**Comment by whyman** [comment]
@whorfin Nice to hear from you again :smile: 

I did have patch to build a transaction per file which should speed it up a bit, would you be able to test it if I can find/write it again?

**Comment by whyman** [comment]
@whorfin Can you try the patch from https://github.com/gerbera/gerbera/pull/1278?

**Comment by whorfin** [comment]
@KarlStraussberger Ah ok, now I understand, I thought you were indicating that had been fixed at some point.  Since gerbera handles shutdown events, and under normal circumstances exits with a pleasant "Have a nice day", wouldn't that be an opportunity to checkpoint scan status so that it might be resumed?

@whyman Thanks!  I will set about pulling/building/testing your sql-transactions branch and let you know how it goes.

**Comment by KarlStraussberger** [comment]
Nice idea, but there are several tasks running in parallel and we would need to store the whole `lastModified` dictionary from autoscan just in case the first run is interrupted.

**Comment by whorfin** [comment]
Run against the "post migration" backup of the database, CPU once again hammered - 100% of one core
Plus side - things did seem snappier, and `8player` was perfectly usable.  `foobar2000` still not usable at all - is there any chance that a non-chunked "give me all entries" request would block or never complete while database is still scanning in such a state?

Now installed and running to rebuild the freshly-built-in-1.7.0 database which, worryingly, hadn't yet rebuilt overnight.  CPU now higher for this build than it was previously.
`foobar2000` still unusable, `8player` OK

Additionally, I tested with VLC on desktop, and that was OK.  There appears to be some pathological interaction between the post 1.6.1 versions and foobar2000 on iOS, which I'm at least hoping will resolve when the scan is complete.  foobar2000 on Windows is fine.

I'll continue to let you know how it goes

**Comment by KarlStraussberger** [comment]
Thanks for supporting us. 100% is fine (I observed the same). My 100K media files (25% MP3 and 75% JPG) took about 5 hours, as far as I remember. At least less than a day!

**Comment by whorfin** [comment]
One more client info update:
`VLC` on iOS - works fine
`BubbleUPnP` on Android - works fine
`foobar2000` on Android - same issue as with iOS, at least while scan is still happening, unusable

**Comment by whorfin** [comment]
Almost a day in, scan seems to be still happening - occasional taglib messages, the database grows very slowly.  While DLNA access seems perhaps a bit faster with the WAL changes, it's still very slow and `foobar2000` remains unusable.  The changes made for this round of testing [UPNP bump and new branch] haven't broken anything, but they do not seem to have addressed whatever the fundamental issue is.
Gerbera is using 100% of one core but is not I/O limited, either in the database or the media files.  Much CPU is being spent on ... something ... 
I will continue to update as things change, which hopefully some sanity returns after the scan completes, whenever that might be

**Comment by KarlStraussberger** [comment]
I got it. Not sure that it solves the foobar hang but import is as fast as ever. PR follows.

**Comment by whorfin** [comment]
Fantastic!

Doing a little process spying, i am seeing what appear to be attempted error messages which are not showing in the logs, ie writes directly to stderr which are discarded due to daemonization
for example:
```
/11:     0.0001 write(2, 0x008CDA84, 37)                        = 37
/11:       l i b u p n p   T h r e a d P o o l A d d   t o o   m a n y   j
/11:       o b s :
/29667:  0.0000 lwp_park(0x00000000, 0)                         = 0
/11:     0.0001 write(2, " 1 0 0\n", 4)                         = 4
```

**Comment by whyman** [comment]
The dreaded `too many jobs`.

When you see it around the internet its usually because of a non-responsive device, causing threads to hang waiting for response.

**Comment by whorfin** [comment]
This was with the 1.16-candidate master branch of libpupnp
FWIW, along with eventually fixing it, it seems there's a problem now in that this message vanishes as it is not logged properly in the way gerbera messages are.
I'll try rebuilding with --disable-blocking_tcp_connections
Thanks for the fixes for database, will give a try!

**Comment by whorfin** [comment]
OK, database rebuilding away again
We'll see if it can rebuild without an overnight!
Much less taglib spewage, so thanks!, though now there's some unhelpful info messages (see #1277)
`foobar2000` still useless, though now if the database rebuild completes in reasonable time, we'll see if that sorts itself once database is complete

**Comment by whorfin** [comment]
while it's churning...
a quick profile window shows 30% CPU spent in `sqlite3VdbeExec`; next closest is `sqlite3VdbeRecordCompareWithSkip` at 5%

Sorting it all by library, looks like this:
```
LIBRARY                                                 COUNT   PCNT
libiconv.so                                                 1   0.0%
ld.so                                                       2   0.0%
libtag.so                                                  13   0.2%
libumem.so                                                101   1.3%
libstdc                                                   101   1.3%
gerbera                                                   140   1.8%
libmagic.so                                               153   2.0%
libc.so                                                   858  11.3%
libsqlite3.so                                            6256  82.0%
```
so, basically hammering the database at this point; FWIW

**Comment by KarlStraussberger** [comment]
I expected that, because all operations end up in database.

The only way to reduce traffic is to cache CdsContainer objects because they are searched for each file. If we assume that the tree is populated in a structured way, that at least a container contains some 10 entries and is 4 layers deep we save 10^4 reads for the children of the root container. On the other hand, runtime memory goes up.

**Comment by whorfin** [comment]
gerbera has been running for 16 hours on new build, 100% cpu on one core
`fooba2000` (and now `vlc` too) both hanging on the largest directory per above, `8player` still ok
no log entries since midnight but process trace shows it is still scanning
tracing syscall shows an extraordinary number of seek/read pairs for the database file, eventually a stat/open of a music file, then some locks/unlocks and upnp chatting
although the filesystem can certainly handle it, there is a surprising amount of duplicate access to files - 19 stats per music file, for example

On the chance this is helpful, here's the system calls from the first stat of a music file [during this rebuild scan] to just before the stat of the next
[trace.txt](https://github.com/gerbera/gerbera/files/6092166/trace.txt)

**Comment by KarlStraussberger** [comment]
Access to media files is completely encapsulated in libraries like taglib, so this is  out of our reach.

I did a little experiment with caching containers and was able to gain another 15%: 135 minutes instead of 155 for my 100k files.

**Comment by KarlStraussberger** [comment]
As far is I can see `foobar2000` tries to load all details from a container. It browses the list and the requests details on **each** entry - from the speed it loads normal containers: all in parallel. That's the reason for the lock.

Somehow we need to block thIs kind of DoS calls.

**Comment by whorfin** [comment]
FWIW during this most recent round that regular ol' Desktop VLC also hangs when doing the same operation, ie  "PC Directory" access to a folder with many entries; in my case, there are no media files in the directory that makes it choke, there are 925 subdirectories.
VLC spins for a while, then gives up and accessess the next sibling directory

**Comment by KarlStraussberger** [comment]
Did similar checks already, vlc and foobar, about 500 entries still work.

The problem seems to lie between those two numbers.

**Comment by whyman** [comment]
I think its most likely that the libupnp miniserver itself is getting overwhelmed rather than our code

**Comment by KarlStraussberger** [comment]
Let me put together that PR and make a 1.7.1 then.

**Comment by whorfin** [comment]
> I think its most likely that the libupnp miniserver itself is getting overwhelmed rather than our code
...except that in 1.6.1 with a scanned database, `foobar2000` had no problem with those directories.  Doesn't that point to gerbera?
36 hours in with the latest build and scanning is still slowly proceeding.

**Comment by KarlStraussberger** [comment]
> ...except that in 1.6.1 with a scanned database, `foobar2000` had no problem with those directories. Doesn't that point to gerbera?

In order to make sure it's gerbera: Did you use the same libpupup with 1.6.1?

> 36 hours in with the latest build and scanning is still slowly proceeding.

Thats quite a lot, which import layout do you use? js - builtin - strucutured? What's the hardware platform? My database is about 620MB.

**Comment by whorfin** [comment]
Yes, i used the same libpupnp as with 1.6.1 before moving to master as suggested; no change
I'm using "builtin" layout and js is disabled
The system is a four-core AMD64 @ 2.5GHz with 8G memory; it's not a router or dinky rpi

**Comment by KarlStraussberger** [comment]
Even more confusing. There is no critical change in builtin layout at all. Must be a completely different reason, then.

**Comment by whorfin** [comment]
> Access to media files is completely encapsulated in libraries like taglib, so this is out of our reach.


I did an experimental build with taglib disabled, and I see 18 stats instead of 19
So I do not believe this is due to taglib

**Comment by whorfin** [comment]
It was worth a shot, and worth fixing, but i don't think that was it - i patched by hand and didn't see noticeable improvement
I added timestamps to the trace, and it confirms something is bad - during the "i see 19 stats" of a music file, i see 4-5 seconds pass.  5 seconds per file is ... not good

and then i did the same with 1.6.1; approximately .1 seconds for a file, which is ... better
[trace-1.6.1.txt](https://github.com/gerbera/gerbera/files/6097749/trace-1.6.1.txt)
[trace-recent.txt](https://github.com/gerbera/gerbera/files/6097750/trace-recent.txt)

**Comment by whyman** [comment]
That is indeed a boatload of seeks! Do you have debug logging enabled? Should allow us to see the queries

**Comment by whorfin** [comment]
I don't [yet], the main difference seems to be TOMBDEBUG?  I ran with --debug, which seems to otherwise do the same thing vis spdlog; is this what you are looking for?

i've attached debug output plus a trace run at the same time; since i don't know exactly how things match up, i'm bracketing the first time a track appears in each - 
wob_microwave_1.mp3 through wob_microwave_3.mp3 and ended first time wob_microwave_4.mp3 appears; so there should at least be complete info for wob_microwave_3.mp3
[trace-with-debug-trimmed.txt](https://github.com/gerbera/gerbera/files/6097918/trace-with-debug-trimmed.txt)
[debug-log-trimmed.txt](https://github.com/gerbera/gerbera/files/6097920/debug-log-trimmed.txt)

**Comment by whorfin** [comment]
oh ... looks like if BufferedIOHandler is involved, which at first blush i thought was just for sockets, TOMBDEBUG might be helpful
ok, i will make a WITH_DEBUG build and repeat the process above...

**Comment by KarlStraussberger** [comment]
Unfortunately #1285 did not break the spell.

Just to do anything, can you share your compile-info and relevant parts from config.xml (only new or changed).

**Comment by whyman** [comment]
So I had a closer look at @whorfin's debug log and what one thing I noticed was that `_purgeEmptyContainers` SQL takes 2 seconds:

```
2021-03-07 12:26:35.249  debug: [sql_database.cc:1459] _purgeEmptyContainers(): start upnp: 1821913,9,108237,1348556,922598,21,24,80158,29,108247; ui: 
2021-03-07 12:26:35.249  debug: [sql_database.cc:1501] _purgeEmptyContainers(): upnp-sql: SELECT "a"."id", COUNT("b"."parent_id"),"a"."parent_id","a"."flags" FROM "mt_cds_object" "a" LEFT JOIN "mt_cds_object" "b" ON "a"."id"="b"."parent_id" WHERE "a"."object_type"=1 AND "a"."id" IN (1821913,9,108237,1348556,922598,21,24,80158,29,108247) GROUP BY a.id
2021-03-07 12:26:35.249  debug: [sqlite_database.cc:351] select(): Adding select to Queue: SELECT "a"."id", COUNT("b"."parent_id"),"a"."parent_id","a"."flags" FROM "mt_cds_object" "a" LEFT JOIN "mt_cds_object" "b" ON "a"."id"="b"."parent_id" WHERE "a"."object_type"=1 AND "a"."id" IN (1821913,9,108237,1348556,922598,21,24,80158,29,108247) GROUP BY a.id
2021-03-07 12:26:37.173  debug: [update_manager.cc:206] threadProc(): sending updates...
```

I think in general the "delete stuff" path is pretty unoptimised as it wasn't in the import path until recently.

**Comment by KarlStraussberger** [comment]
Looks like a **`lock`** which is bricked. I found stuff like this:

```
2021-03-07 12:26:06.571  debug: [sqlite_database.cc:368] exec(): Adding query to Queue: UPDATE "mt_cds_object" SET "update_id"="update_id" + 1 WHERE "id" IN (108247,29)
2021-03-07 12:26:07.526  debug: [sqlite_database.cc:619] run(): UPDATE "mt_cds_object" SET "update_id"="update_id" + 1 WHERE "id" IN (108247,29)
```

There is almost a second between adding the query and running it.

**Comment by whorfin** [comment]
> Just to do anything, can you share your compile-info and relevant parts from config.xml (only new or changed).

Latest cmake flags:
```
-DWITH_MAGIC=0 -DWITH_MYSQL=0 -DWITH_CURL=0 -DWITH_JS=0 -DWITH_TAGLIB=0 -DWITH_AVCODEC=0 -DWITH_EXIF=0 -DWITH_LASTFM=0 -DWITH_SYSTEMD=0 -DWITH_INOTIFY=0 -DWITH_MATROSKA=0 -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=NO 
```
Not much fancy in config, i'll just attach it for surety.  Only relatively new change is changing _alive_ from 1800 to 180 while debugging PIMD.  The rest is basic - no transcoding (there's some old PS3 cruft but disabled)), shenanigans to disable playlists, some mime tweaks.

[config-redacted.xml.txt](https://github.com/gerbera/gerbera/files/6103061/config-redacted.xml.txt)

**Comment by KarlStraussberger** [comment]
I found a place where loads of time is spent

```
2021-03-07 12:26:51.301  debug: [sqlite_database.cc:351] select(): Adding select to Queue: SELECT COUNT(*) FROM "mt_cds_object" WHERE "parent_id"=24
2021-03-07 12:26:52.110  debug: [update_manager.cc:142] containerChanged(): containerChanged. id: 24, signal: false
```

After each `addObject` the number of items is checked to send an update: https://github.com/gerbera/gerbera/blob/9947a456b2c87f5198e92ec0c59dff44bb097ca2/src/content/content_manager.cc#L1025 

As a first attempt you may include the following lines in `/usr/local/share/gerbera/sqlite3.sql` before the final `COMMIT`

```
CREATE INDEX mt_cds_object_id ON mt_cds_object(id);
CREATE INDEX mt_cds_object_parent_id2 ON mt_cds_object(parent_id);
```

That should speed up the select and we do not have to change anything else. It may also fix your browsing issue, because parent_id is used in some browse request also.

**Comment by whorfin** [comment]
I see approximately 6 seconds per audio track; that's an improvement over the 9 previous although of course the database is in a slightly different state
Good "chipping away" progress
Let me know if you'd like fancier diagnostics; this was just quick timestamp tracing

On a related note - if you're considering adding additional indices, way back when https://github.com/gerbera/gerbera/pull/183
I noted:
```
CREATE INDEX mt_cds_object_parent_type ON mt_cds_object(parent_id,object_type);
would speed some queries by as much as 3x; yes these fields are present in an existing index, but the scheduler is affected in a way that can matter.
```

**Comment by KarlStraussberger** [comment]
I'm sorry that index did slow down on my side from two to almost four hours. But we learned from your test, that we're close to an issue. So I will try to get rid of the childcount thing.

**Comment by KarlStraussberger** [comment]
Which version of sqlite do you have? Seems to affect the way it handles count and indixes.

**Comment by KarlStraussberger** [comment]
@whorfin Do you mind running that "final" test? You have to revert your changes to the sqlite3.sql before. Thank you! :100:

**Comment by whorfin** [comment]
@KarlStraussberger I had sqlite3-3.31.1 
Upgrading to sqlite3-3.34.1 now
I will rebuild with the patched content_manager.cc and revert the sqlite3.sql change
...isn't that index persistent, though, after being created?

**Comment by KarlStraussberger** [comment]
yes, you need to drop the database, also

**Comment by KarlStraussberger** [comment]
If you remove the sqlite file, everything is gone. Good and bad.

**Comment by whorfin** [comment]
I'm sorry, I don't follow.
You want me to test with complete database rebuild?

**Comment by whorfin** [comment]
...i have enough other things to do and my library updates broke spdlog so i will get to this later

**Comment by KarlStraussberger** [comment]
> I'm sorry, I don't follow.
> You want me to test with complete database rebuild?

Yes, I thought you did that each time.

You'd - best clone from my branch https://github.com/KarlStraussberger/gerbera/tree/use-cache-for-add

**Comment by KarlStraussberger** [comment]
Updating an incomplete database can produce results which are hard to understand. If you build from scratch, we see typical behaviour. Anyway, speed effects become visible as your database grows. That's why indexes did not work - they require rebuild after each insert. They are good for select and update.

**Comment by whorfin** [comment]
OK, operating on existing database, same results - 6 seconds per file
Starting from scratch, crash immediately after "info: Loading 0 configuration items from database"
```
#0  0xfffffc7fee16d124 in mutex_lock_impl () from /lib/64/libc.so.1
#1  0xfffffc7fee16d273 in pthread_mutex_lock () from /lib/64/libc.so.1
#2  0x00000000005a28a8 in UpdateManager::containerChanged(int, int) ()
#3  0x000000000058004e in ContentManager::ensurePathExistence(std::filesystem::__cxx11::path) ()
#4  0x00000000005862ca in ContentManager::ContentManager(std::shared_ptr<Context> const&, std::shared_ptr<Server> const&, std::shared_ptr<Timer>) ()
#5  0x00000000005d2039 in Server::init() ()
#6  0x0000000000507221 in main ()
```
which led to infinite spew if i attempted to restart:
```
2021-03-09 12:37:23  error: [/home/whorfin/gerbera/gerbera-master-9947a45/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-09 12:37:23  error: [/home/whorfin/gerbera/gerbera-master-9947a45/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
```

**Comment by KarlStraussberger** [comment]
Now we're back on my original idea - some lock or mutex. What compile are you using. Any uncommon library or so. I feel lost.

**Comment by whorfin** [comment]
gcc9
Note that this particular routine works fine in the case where the database already exists, so it is not simply use of locks but something about their states, contention, intitializatin ... or something
locks are the worst

**Comment by whorfin** [comment]
Because i updated libsql3 which pulled in other new libraries, i rebuilt 1.6.1 with exactly the same toolchain and cmake arguments, 
and it also now crashes when trying to create the database
wonderful...
this will take some time to recover from

**Comment by KarlStraussberger** [comment]
Maybe you call pull the changes to my branch and see how it behaves - after removing the broken db file. It may even be possible that the lock in updatemanager created some sort of shift in locking and unlocking.

**Comment by whorfin** [comment]
The behavior reported _was_ with a fresh start - the bogus method crash was always on a fresh start, not when using an existing database.
Just pulled down your updated branch, rebuilt, tried again
gerbera.db removed
run gerbera
```
2021-03-10 08:47:08   info: Checking configuration...
2021-03-10 08:47:08   info: Configuration check succeeded.
2021-03-10 08:47:08 warning: Sqlite3 database seems to be corrupt or doesn't exist yet.
2021-03-10 08:47:08   info: No sqlite3 backup is available or backup is corrupt. Automatically creating new database file...
2021-03-10 08:47:08   info: database created successfully.
2021-03-10 08:47:08   info: About to migrate metadata from mt_cds_object to mt_metadata
2021-03-10 08:47:08   info: No data will be removed from mt_cds_object
2021-03-10 08:47:08   info: Migrated metadata - object count: 0
2021-03-10 08:47:08   info: Loading 0 configuration items from database
2021-03-10 08:47:08   info: Loading autoscan location: /tank/whorfin/Audio; recursive: true, last_modified: unset
2021-03-10 08:47:08   info: Initialising libupnp with interface: 'e1000g0', port: 59666
2021-03-10 08:47:08   info: Initialized port: 59666
2021-03-10 08:47:08   info: Server bound to: <local ip>
2021-03-10 08:47:08   info: The Web UI can be reached by following this link: http://<localip>:59666/
2021-03-10 08:47:08  error: Failed to stat /tank/whorfin/Audio/Synth, Permission denied
[...two more permission denied subdirs - this is intentional]
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:08  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
2021-03-10 08:47:09  error: [/home/whorfin/gerbera/gerbera-use-cache-for-add/src/database/sql_database.cc:351] _addUpdateObject Error: tried to create or update an object with an illegal parent id
[etc]
```
Same behavior as observed last night.
I do note the Errors on "illegal parent id" start _after_ some permission denied directories (which are intentional; gerbera is not allowed to share everything in that tree), as if something goes out of sync in the new codepath?
Also, as noted above,  #1107 was the change that introduced the uninitialized member function.  Perhaps other problems were introduced?

**Comment by KarlStraussberger** [comment]
I tried to reconstruct your directory to get the `stat` error, but got different messages. Can you share the relavant tree listing, please.

**Comment by whorfin** [comment]
The stat error is simply from having a directory to which gerbera does not have access
For example:
```
$ ll -d /tank/whorfin/Audio/Synth/
drwx------+ 10 whorfin staff 12 Jan 17 12:56 /tank/whorfin/Audio/Synth/
$ ll -d /tank/whorfin/Audio/Musiq
drwxrwxr-x 926 whorfin media 926 Mar  4 17:20 /tank/whorfin/Audio/Musiq
$ groups gerbera
gerbera : media mxfer
```
gerbera is run as user "gerbera", group "media"
Directories with eg no group nor public access are denied to gerbera, and cause the stat error shown

**Comment by whorfin** [comment]
here's --debug output from start on an empty database to the first bad parent Error, attached
note that your branch reports itself as 1.3.5 i guess because git tags --- there's no way this is 1.3.5 behavior lol...
It looks like it starts trying to add a negative parent id
[debug-db-create.txt](https://github.com/gerbera/gerbera/files/6119023/debug-db-create.txt)

**Comment by KarlStraussberger** [comment]
Thanks, been a dead end. Asumed the missing access caused import to bail out. 

It's been the new caching of containers. Commit to follow.

**Comment by KarlStraussberger** [comment]
> 
> It's been the new caching of containers. Commit to follow.

It's available for another test. Thanks!

**Comment by whorfin** [comment]
Hoorah, that fixes the error on database rebuild
Rebuild is faster than 1.7.0 previously so that's nice, though seemingly not as fast as 1.6.1 - early on in rebuild i see .2 to .5 seconds per music file
Looking at database rebuild, it is pretty dismal - 12 seconds per file in some early tests
on main database where it is rescanning aiff right now, about 6 seconds per file

Attaching an example where you can see an absurd amount of read/seeking and what looks like very weird threading at some points - read/lock/seek/lock/unlock/unlock all mingled together
spell still not broken
[trace-newest-restart.txt](https://github.com/gerbera/gerbera/files/6125543/trace-newest-restart.txt)

**Comment by KarlStraussberger** [comment]
I've scanned the code for all the stats
- scanning directory (1)
- metadatahandler (2)
- TaglibHandler (?)
- FfmpegHandler (?)

Can you share a new corresponding debug printout? It's easier to follow the import process and action from that.

Is it possible that some active client is trying to browse or stream while import is going on, because all the reads and seeks do not belong to the scan (it's file id 3 or 4 mostly - might belong to sqlite).

**Comment by whorfin** [comment]
I don't believe there are any active clients; using a larger data display window on trace shows fid 3 and 4 point to them being most likely .db and .wal file, ie sqlite3, which would be consistent with file id usage - 0,1,2 being stdin/out/err  (i don't see anything looking like track content there)
No FFmpegHandler in use, and per previous posts, disabling taglib only cut down the number of stats by 1.

here's a quick snip of a relatively fast, in that i've seen 12 seconds and above per track, action when adding one track
[debug-log-trimmed.txt](https://github.com/gerbera/gerbera/files/6132249/debug-log-trimmed.txt)

Even the half-second-per-track on an empty database start, sent above, is too slow.  That's almost four days to import.  I can rebuild my Kodi database in hours over a slow network link to a different old machine  (gerbera is on the same machine as the filesystem).  This iOS app
https://www.stratospherix.com/products/musicstreamer/
does a really nice feature of two-pass scanning - first pass gets minimal information, and scans my entire library over wifi in less than twenty minutes.  After first pass, all music is available for play; second pass fetches full information and album art, and takes a few hours - it is easily interrupted/restarted with no issues
Just offering some comparables to show what kind of performance is possible, and really hoping _gerbera_ can get back to previous performance, and hopefully do even better.

**Comment by KarlStraussberger** [comment]
I see that there is weird some issue, we already try to resolve to get gerbera fast again. We might consider some architectural changes in the future.

At the moment I'm about to introduce `fs::directory_entry` instead of the C functions. The first changes brought another 10%. Looking at where we started that's almost 40% altogether. I hope that by cleaning things up we resolve the mutex issue on your system.

**Comment by KarlStraussberger** [comment]
I created the new PR (the other branch was merged into master): https://github.com/gerbera/gerbera/pull/1312

It should reduce the stats, because of `fs::directory_entry`. There is no change to the mutexes yet.

**Comment by whorfin** [comment]
Running with no database to start, i see 21 stat() system calls per mp3, about .32 seconds elapsed per file early in the process; this increased not decreased
Running rebuilding, approx 10.5 secs/file
[debug-restart-trimmed.txt](https://github.com/gerbera/gerbera/files/6144139/debug-restart-trimmed.txt)
[trace-restart-snipped.txt](https://github.com/gerbera/gerbera/files/6144141/trace-restart-snipped.txt)

**Comment by KarlStraussberger** [comment]
I doubt that rebuilding the database will be successful. It's broken, because of not being completed in the first place. In such a rare situation we have to remove each item and re-add it (there's no fast-lane update since the beginnings of mediatomb - that's something on my list to fix). I'd give up that Db - building it from scratch should be our focus.

Speaking of the creation, it slows down the process if no tags (artist etc.) can be read from the file. Looking at the time spent in the stats they don't seem to be the issue. My scan ran on 31000 files in 60 mins with builtin layout, that's approx. 0.12 sec per file.

**Comment by whorfin** [comment]
Hmm, ok then, i did not realize there was that lurking trap; is there a log message when the db is in a state where restarting is safe?
A quick estimate right now looks like 17-18 hour estimate given some stat timing peeking

Also - 
Restarting with no database, i get the message about creating new database, and that "database created successfully"
The first and so far subsequent messages of "Failed to stat" for the directories without permission look like this:
```
2021-03-15 16:33:01  error: _rescanDirectory: Failed to stat /tank/whorfin/Audio/Synth, Invalid argument
```
Really?  _rescanDirectory() doesn't sound like the place it should be for a new database, yes?
(also, "Invalid argument" seems wrong; i used to see "Permission denied" which is correct)

**Comment by KarlStraussberger** [comment]
> Restarting with no database, i get the message about creating new database, and that "database created successfully"
> The first and so far subsequent messages of "Failed to stat" for the directories without permission look like this:
> ```
> 2021-03-15 16:33:01  error: _rescanDirectory: Failed to stat /tank/whorfin/Audio/Synth, Invalid argument
> ```
> Really?  _rescanDirectory() doesn't sound like the place it should be for a new database, yes?
> (also, "Invalid argument" seems wrong; i used to see "Permission denied" which is correct)
> 
> 
Lazyness on my side. Left out error handling when first reading a directory in hope for the remaining code to catch. It did, but the message is worth adding the handling in the right place.

**Comment by whorfin** [comment]
huh
yes it's weird, two more directories had the same message
two more had "addRecursive: Failed to stat" with correct error code (Permission denied)
seems to be the difference between top-level and subdir, which might be what you are referring to

**Comment by KarlStraussberger** [comment]
Yes. The main dir is the entry point and the subdirs are handled with.

Added a PR to handle error where they occur.

**Comment by neheb** [comment]
Just hit this. master is unusable on my NAS. Didn't use to be the case.

**Comment by KarlStraussberger** [comment]
> 
> 
> Just hit this. master is unusable on my NAS. Didn't use to be the case.

Already fixed.

**Comment by whorfin** [comment]
sigh
blew up during the scan; my service system restarted it but as you have said and we have demonstrated, that's just useless because rescans don't work properly until the first is completed
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  stoi
```

**Comment by KarlStraussberger** [comment]
if you pull again, the error should be gone.

**Comment by whorfin** [comment]
I will have to return to this later, but i've already found a bug, which happened on restart
PR #1318 for the fix, which should also help things crash faster on Solaris, as there was an ancient gross hack which allowed a nullptr dereference to succeed.

**Comment by whorfin** [comment]
Also - @KarlStraussberger 
after most recent pull, and despite your recent PRs to clean this up, I still get bogus error messages
```
2021-03-16 11:40:10  error: _rescanDirectory: Failed to stat /tank/whorfin/Audio/Synth, Invalid argument
2021-03-16 11:40:10  error: _rescanDirectory: Failed to stat /tank/whorfin/Audio/ipwn, Invalid argument
```
etc; /tank/whorfin/Audio is the timed autoscan entry, so those are one-level-down
two and further levels down get the right message (well they did before the recent pull, haven't gotten there yet in the new scan) -- i'd previously seen 
```
addRecursive: Failed to stat /tank/whorfin/Audio/foo/bar, Permission denied"
```
Note that the error codes are delivered completely differently between those two bits of code, only one of which works properly

**Comment by whorfin** [comment]
Finally!
Complete rebuild in about 6 hours
`mobile foobar2000` works just fine again, very snappy
presumably a rebuild of an unfinished database would still be awful
at least until that situation is sorted, it could be argued that a simple database status should be tracked, and a gerbera restart cause complete rebuild until stability is ensured

**Comment by whyman** [comment]
@whorfin great to hear! I think there is still some room for improvement, but that sounds much more reasonable.

**Comment by whorfin** [comment]
I pulled the absolute latest (f18d511), and the _rescanDirectory messages are not only not fixed, things got worse, and valid directories are rejected with a "Failed to iterate" message:
```
2021-03-17 13:31:50  error: _rescanDirectory: Cannot access entry /tank/whorfin/Audio/Synth, Invalid argument  # should be Permission denied
2021-03-17 13:31:50  error: _rescanDirectory: Failed to iterate /tank/whorfin/Audio/Wumpscutt-BarbsRemixLoops, Invalid argument # bogus - permissions ok and used to work
2021-03-17 13:31:50  error: _rescanDirectory: Failed to iterate /tank/whorfin/Audio/Sounds, Invalid argum # similarly incorrect
```
i can hop this to a different issue report but since so many things are changing at once, and there's lots of context here, figured this to be hopefully appropriate

**Comment by KarlStraussberger** [comment]
I'm fine with putting the discussion.

Obviously, I need to refine my local test directory to reflect your tree. Hope, I'm getting all combinations sorted out.

**Comment by KarlStraussberger** [comment]
I prepared a structure with protected dirs and files. Now, I get the "**Permission denied**" for directories again but `directory_entry` has no option the check access to files. So the errors on protected files come from the metadata handlers which actually read them.

**Comment by whorfin** [comment]
I'm not sure I follow
Do you have a patch to try?
As reported above, the bogus errors I am complaining about are directory permissions, improperly reported as "Invalid argument"
If they happen to a direct child of the timed autoscan directory, they are improperly reported as _rescanDirectory on the first scan, and improperly reported as 'Invalid argument'
If they happen to an indirect child [ie child of a child of timed autoscan directory] i see the error as reported by "addRecursive" which is correct, but with "Invalid argument" instead of the correct "Permission denied"
The recent change broke the addRecursive message, and has something to do with the "ec' object versus the previous strerror(errno) method
The other very unwanted behavior is that now some of these "children of direct child" cannot access entry are just wrong, those directories used to be scanned and in the new version are bogusly skipped; the message is slightly different - _rescanDirectory: Failed to iterate for the bogus child-of-child message,  reported as "Invalid argument"
  note - some of these "Failed to iterate"  messages are indeed for protected directories, some are wrong (ie the directory is not protected, and caused no issues previous to these changes)
again, all reported above, just trying to be clear

none of my testing so far involves protected files - protected files are never seen by gerbera as their enclosing directory is also protected

**Comment by KarlStraussberger** [comment]
The PR is already merged into master.

I changed the handling completely so that on my side it worked as expected. The message about the zip file is caused by too many quick fixes while doing other things.

**Comment by whorfin** [comment]
oh
i see what's happening; the interator is being called and used incorrectly
PR soon

**Comment by whorfin** [comment]
#1328

**Comment by whorfin** [comment]
Here's a new fun permissions problem
With database rescan complete, I can now try server restart
There's no "i'm done scanning" message so i am just guessing from CPU load and a trace, which seems to show files not being touched
Rescan was very fast this time which is great
What's not great:

Scan your music library
Stop gerbera when done scanning
Gone "prune" a branch by locking it off with permissions (as opposed to deleting) at the directory level
Start gerbera again and let it Rescan
_rescanDirectory will correctly say "Failed to iterate" (or addRecursive, depending on dir level, but in my tests this was at top level and _rescanDirectory is what was hit; this is extrra confusing because addRecursive() now incorrectly logs itself as _rescanDirectory at line 874
But as can be seen from the code, which simply returns in this case, the directory is _not_ pruned from the database. 
So gerbera advertises directories [under PC Directory] and tracks which it _can not serve_; they show up as unknown type in eg 8player

Fixing this would require "_rescanDirectory" not being called the first time through as it is now

**Comment by whorfin** [comment]
also @KarlStraussberger 
>This brings back the permission denied messages from the C code.

Sadly no, still "Invalid argument" on my machine, but i like seeing _something_ indicating permissions issues, as sometimes i need to fix things
FWIW, as i mentioned above, ec.message() in current code doesn't  produce "Permission denied" as the C code did
the semantics are fairly different now, and note  is_directory returns false with std::filesystem directory_entry when permissions are denied; as previously mentioned, exists() says false as currently called
this is g++ 9 on OmniOS/Illumos
not a dealbreaker of course, but annoyingly uninformative

**Comment by KarlStraussberger** [comment]
I moved that to a new issue, because I doubt that case was handled correctly in previous versions. The message was just a leftover to be removed

**Comment by whorfin** [comment]
Update works fantastically, thank you!

**Comment by joerg-krause** [comment]
> Update works fantastically, thank you!

Great progress! Many thanks to all!

How about a bugfix release 1.7.1 at this point?

**Comment by whyman** [comment]
I think there are probably enough changes since `1.7.0` for `1.8.0`.

Lets close this off and think about tracking a new release.

Thanks to everyone in the thread and especially for you patience @whorfin!
